### PR TITLE
feat(docker): add support for Basic auth

### LINKED ISSUE #3059: Support Artifactory docker authentication
**What would you like Renovate to be able to do?**
Currently, we host Renovate internally and run it on a slew of projects. Some point to the external docker registry, and others can point to our internal mirror/enterprise registry. Our expectation would be that in all cases, the bot knows how to pin and upgrade SHAs as well as different containers. Right now, I get `406` errors when trying to update CircleCI configs that contain an internal registry endpoint. To my understanding, it is an all or nothing process. Either everyone uses the registry or no one does.

**Describe the solution you'd like**
I'd like to be able to provide recognition for our internal registry url (something like `docker.enterprise.com`) as well as the necessary credentials (username and password) to be able to look up containers. I'd also like the current ability of pinning containers in CI configs to transfer over when using an internal mirror. However, if I switch to a repo using a public mirror, the functionality should remain the same as it is today.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by derkoe**
This adds basic auth support to the docker registry datasource. This enables to us use private Docker registries like Nexus and Artifactory.

Closes #3059

### CODE REVIEW

**Comment by rarkins** [comment]
@derkoe thanks very much! Can you describe quickly how much testing you've done against real repos and registries? e.g. Nexus or Artifactory, and Docker Hub too? Tested against both tags as well as digest updates?

File: lib/datasource/docker/index.js [hunk 1]
```diff
@@ -59,6 +57,15 @@ async function getAuthHeaders(registry, repository) {
       );
       opts.headers = { Authorization: `Basic ${auth}` };
     }
+
+    if (authenticateHeader.scheme.toUpperCase() === 'BASIC') {
+      logger.debug(`Using Basic auth for docker registry ${repository}`);
+      await got(apiCheckUrl, opts);
```
Comment (rarkins) [inline lib/datasource/docker/index.js]: @derkoe I know this is an old PR, but I was looking at this line today and confused as to what it's achieving. It's calling `got` but not saving the response.

Comment (derkoe) [reply lib/datasource/docker/index.js]: OK - I think that line can be removed.

**Comment by rarkins** [comment]
Actually I think I figured out the reason. I think you were attempting to test early if the auth works or not, as it will then throw an error of a 4xx is returned

**Comment by derkoe** [comment]
Yes that might be it - so that it won't fail later on the Docker pull itself

**Pull request merged**